File: java/src/main/java/it/giannibombelli/wsc2026/payment/domain/policies/PaymentCharging.java
package it.giannibombelli.wsc2026.payment.domain.policies;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.payment.domain.events.TransactionStarted;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentProvider;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentProviderResult;

import static java.util.Objects.requireNonNull;

public final class PaymentCharging {

    private final PaymentProvider provider;

    public PaymentCharging(PaymentProvider provider) {
        this.provider = requireNonNull(provider);
    }

    public PaymentProviderResult charge(TransactionStarted event) {
        Require.requireArgument(event, "event");
        return provider.process(
            event.aggregateId().value(),
            event.transactionId().value(),
            event.amount()
        );
    }
}

File: typescript/src/payment/application/services/paymentCharging.ts
// Policy che invoca il provider di pagamento a fronte di una TransactionStarted.

import { TransactionStarted } from '../../domain/events/transactionStarted.js';
import type { PaymentProvider } from '../../domain/ports/paymentProvider.js';
import type { PaymentProviderResult } from '../../domain/ports/paymentProviderResult.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class PaymentCharging {
  private readonly _provider: PaymentProvider;

  constructor(provider: PaymentProvider) {
    requireDependency(provider, "provider");
    this._provider = provider;
  }

  charge(event: TransactionStarted): PaymentProviderResult {
    requireArgument(event, 'event');
    return this._provider.process(event.aggregateId.value, event.transactionId.value, event.amount);
  }
}
